torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,638 @@
+                from a call to :meth:`state_dict`
+        """
+
+        # Check whether we got a local or global dict
+        if state_dict["local_state_dict"]:
+            self.load_local_state_dict(state_dict)
+        else:
+            # Get this optimizer's param_groups shard
+            param_groups = state_dict["param_groups"][
+                state_dict["partition"][self.rank][0] : state_dict["partition"][self.rank][1]
+            ]
+            # Dispatch this rank's state dictionary to the wrapped shard optimizer
+            self.load_local_state_dict({"state": state_dict["state"][self.rank], "param_groups": param_groups})
+
+    def add_param_group(self, param_group: Dict) -> None:
+        """Add a param group to the :class:`Optimizer` s ``param_groups``.
+
+        This can be useful when fine tuning a pre-trained network as frozen layers can be made
+        trainable and added to the :class:`Optimizer` as training progresses.
+
+        Arguments:
+            param_group (dict): Specifies what Tensors should be optimized along with group
+            specific optimization options
+
+        .. warning: This handles updating the shards on all partitions, but needs to be called on all ranks.
+        """
+
+        super().add_param_group(param_group)
+        if not self.in_super_constructor:
+            # Force a re-partitioning
+            self._partition_parameters.clear()
+            self._per_device_params.clear()
+            self._param_rank.clear()
+
+            param_groups = self.partition_parameters()[self.rank]
+            if len(param_groups) == len(self.optim.param_groups) + 1:
+                self.optim.add_param_group(param_groups[-1])
+
+    ##########################
+    # Private helper functions
+    ##########################
+
+    def _broadcast_params(self) -> None:
+        """Helper function to broadcast all the parameters from a given device"""
+
+        with torch.no_grad():
+            # All the params on this device (inc all ranks)
+            for (
+                device,
+                device_params,
+            ), bucket_list in zip(self.per_device_params.items(), self._buckets.values()):
+
+                # Go through all the ranks and broadcast the new parameters, bucket the smallest ones
+                for (src_rank, params), bucket in zip(enumerate(device_params), bucket_list):
+
+                    global_src_rank = self.get_global_rank(self.group, src_rank)
+
+                    for param in params:
+                        # Bucket broadcast
+                        if self.should_bucket_param[param]:
+                            assert bucket.append(param), "Bucket overflow: max %s - current %s - adding %s" % (
+                                bucket.max_size,
+                                bucket.current_offset,
+                                param.numel(),
+                            )
```

Comment:
It's a free floating parameter so this could be 25MB, experimentally a very small bucket was faster (with slow interconnect) hence the very small default value, but it's very strange to me also. It's something we discussed with @mrshenli some time ago, I would very much welcome any speed improvements here. 
A very handwavy guess could be that this being in python, the bucketing has a higher cost (extra copies and dispatch) vs cpp for DDP, at least with this implementation. This makes me think that these should probably be tensor views instead?